src/libraries/System.Net.Http/src/System/Net/Http/DiagnosticsHandler.cs

```diff
@@ -58,6 +86,16 @@ internal static bool IsEnabled()
                 throw new ArgumentNullException(nameof(request), SR.net_http_handler_norequest);
             }
 
+            // Since we are reusing the request message instance on redirects, clear any existing headers
+            // Do so before writing DiagnosticListener events as instrumentations use those to inject headers
+            if (request.WasRedirected() && _propagatorFields is HeaderDescriptor[] fields)
+            {
+                foreach (HeaderDescriptor field in fields)
+                {
+                    request.Headers.Remove(field);
```

Comment:
What guarantees that all of the fields the propagator specifies are associated with Headers rather than Content.Headers?

Reply:
Nothing.
Realistically using what we consider content-specific headers for trace propagation sounds like a bad idea though.

Reply:
Is the concern that the propagator would add a header we fail to remove, or that we remove a header the propagator didn't add, or something else?

We added the header earlier using `request.Headers.TryAddWithoutValidation(descriptor, value)`. I'm no expert in this code but it seemed logical that `request.Headers.Remove(field)` would be the symmetrical operation.

Reply:
The concern is that `request.Headers.Remove` won't remove some headers that may only live on `request.Content.Headers`.
Likewise, `request.Headers.TryAddWithoutValidation` would fail to add the same set of headers.

There is a set of headers we consider as [KnownHeaders](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Net.Http/src/System/Net/Http/Headers/KnownHeaders.cs#L34) internally. Any header marked with `HttpHeaderType.Content` could not be added by the propagator unless this code attempts to work with both the request and content headers.

The full list of such headers (as of right now) is [here](https://github.com/microsoft/reverse-proxy/blob/9c6a822f5682c61ddc9d663450979b1c34346285/src/ReverseProxy/Forwarder/RequestUtilities.cs#L89-L99).

I can make the code more permissive, but I don't see a scenario where anyone would want to set any of those headers in a propagator (and if they really wanted to they could look at the carrier object).

Reply:
> I can make the code more permissive

I don't think we need to, as long as we're clear what these things can and can't do.  As written, they can't modify content headers.  If that's ok, great.

Reply:
Agreed with @stephentoub - I am unaware of any propagator that would want to write one of those special headers and I think it is fine that we don't allow it.